### PR TITLE
Update to PyTorch 2.5.1 and Python 3.12 (And fix some things)

### PR DESCRIPTION
**Pull request by RaameshB**
Still working on fixing the actual tests in the repository, but this should work once that is done.

### CODE REVIEW

**Comment by RaameshB** [comment]
There was a little eigenvalue piece within the code that was deprecated, fixed it so the package works with the latest PyTorch.

**Comment by RaameshB** [comment]
This should probably fix the random seed issues that were there earlier. Also, I've modified some of the tests, some need manual checking, which I will do later. They are commented out and have TODOs.

**Comment by RaameshB** [comment]
Gradient descent works now, all I did was pass the ESN hidden states into a linear layer if the learning algorithm was set to 'grad'

**Comment by nschaetti** [comment]
Merged
